src/dbnode/encoding/proto/common.go

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"reflect"
+
+	"github.com/m3db/m3/src/dbnode/encoding/m3tsz"
+
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
+
+type Schema *desc.MessageDescriptor
+
+const (
+	// ~1GiB is an intentionally a very large number to avoid users ever running into any
+	// limitations, but we want some theoretical maximum so that in the case of data / memory
+	// corruption the iterator can avoid panicing due to trying to allocate a massive byte slice
+	// (MAX_UINT64 for example) and return a reasonable error message instead.
+	maxMarshaledProtoMessageSize = 2 << 29
+
+	// maxCustomFieldNum is included for the same rationale as maxMarshaledProtoMessageSize.
+	maxCustomFieldNum = 10000
+)
+
+type customFieldType int
+
+const (
+	// All the protobuf field types that we can perform custom encoding /
+	// compression on will get mapped to one of these types. This prevents
+	// us from having to reference the protobuf type all over the encoder
+	// and iterators and also simplifies the logic because the protobuf
+	// format has several instances of multiple types that we will treat the
+	// same. For example, in our encoding scheme the proto types:
+	// int32, sfixed32, and enums are all are treated as int32s and there
+	// is no reasonm to distinguish between them for the purposes of encoding
+	// and decoding.
+	cNotCustomEncoded customFieldType = iota
+	cSignedInt64
+	cSignedInt32
+	cUnsignedInt64
+	cUnsignedInt32
+	cFloat64
+	cFloat32
+	cBytes
```

Comment:
Any reason for hungarian notation? heh. (Side note https://stackoverflow.com/questions/111933/why-shouldnt-i-use-hungarian-notation)

Reply:
Could you suffix with the word "Field" to avoid collision with say keyword like "float64"?

i.e. 

```go
type customFieldType int

const (
	notCustomField customFieldType = iota
	signedInt64Field
	signedInt32Field
	unsignedInt64Field
	unsignedInt32Field
	float64Field
	float32Field
	bytesField
)
```

Reply:
Does it count as hungarian notation when I'm literally defining new types? I can change this if you feel strongly, but appending `Field` feels just as gross as prepending `c`

Reply:
Well appending a suffix is much more common in our codebase than prefixing with a single character, is what I'm mainly getting at, for solving this type of problem.

Reply:
ok sure